File: myExpenses/src/androidTest/java/org/totschnig/myexpenses/test/provider/AccountTest.java
/*
 * Copyright (C) 2010 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.totschnig.myexpenses.test.provider;

import android.content.ContentUris;
import android.content.ContentValues;
import android.database.Cursor;
import android.net.Uri;

import org.totschnig.myexpenses.model.AccountType;
import org.totschnig.myexpenses.model.Grouping;
import org.totschnig.myexpenses.model.SortDirection;
import org.totschnig.myexpenses.provider.DatabaseConstants;
import org.totschnig.myexpenses.provider.DbUtils;
import org.totschnig.myexpenses.provider.TransactionProvider;
import org.totschnig.myexpenses.testutils.BaseDbTest;

public class AccountTest extends BaseDbTest {

  private final AccountInfo[] TEST_ACCOUNTS = {
      new AccountInfo("Account 0", AccountType.CASH, 0),
      new AccountInfo("Account 1", AccountType.BANK, 100),
      new AccountInfo("Account 2", AccountType.CCARD, -100),
  };

  private void insertData() {

    for (AccountInfo TEST_ACCOUNT : TEST_ACCOUNTS) {
      mDb.insertOrThrow(
          DatabaseConstants.TABLE_ACCOUNTS,
          null,
          TEST_ACCOUNT.getContentValues()
      );
    }
  }

  private long insertAccountWithTwoBudgets() {
    final Grouping grouping = Grouping.MONTH;
    long accountId = mDb.insertOrThrow(
        DatabaseConstants.TABLE_ACCOUNTS,
        null,
        new AccountInfo("Account with 2 budgets", AccountType.CCARD, -100, "EUR", grouping).getContentValues()
    );
    BudgetInfo[] budgets = {
        new BudgetInfo(accountId, "budget 1", "description", 400, grouping),
        new BudgetInfo(accountId, "budget 2", "description", 5000, grouping)
    };
    for (BudgetInfo budgetInfo : budgets) {
      mDb.insertOrThrow(
          DatabaseConstants.TABLE_BUDGETS,
          null,
          budgetInfo.getContentValues()
      );
    }
    return accountId;
  }

  public void testQueriesOnAccountUri() {
    final String[] TEST_PROJECTION = {
        DatabaseConstants.KEY_LABEL, DatabaseConstants.KEY_DESCRIPTION, DatabaseConstants.KEY_CURRENCY
    };

    final String COMMENT_SELECTION = DatabaseConstants.KEY_LABEL + " = " + "?";

    final String SELECTION_COLUMNS =
        COMMENT_SELECTION + " OR " + COMMENT_SELECTION + " OR " + COMMENT_SELECTION;

    final String[] SELECTION_ARGS = {"Account 0", "Account 1", "Account 2"};

    final String SORT_ORDER = DatabaseConstants.KEY_LABEL + " ASC";

    Cursor cursor = getMockContentResolver().query(
        TransactionProvider.ACCOUNTS_URI,
        null,
        null,
        null,
        null
    );
    assert cursor != null;

    assertEquals(0, cursor.getCount());

    insertData();
    cursor.close();

    cursor = getMockContentResolver().query(
        TransactionProvider.ACCOUNTS_URI,
        null,
        null,
        null,
        null
    );
    assert cursor != null;

    assertEquals(TEST_ACCOUNTS.length, cursor.getCount());
    cursor.close();

    Cursor projectionCursor = getMockContentResolver().query(
        TransactionProvider.ACCOUNTS_URI,
        TEST_PROJECTION,
        null,
        null,
        null
    );

    assert projectionCursor != null;
    assertEquals(TEST_PROJECTION.length, projectionCursor.getColumnCount());

    assertEquals(TEST_PROJECTION[0], projectionCursor.getColumnName(0));
    assertEquals(TEST_PROJECTION[1], projectionCursor.getColumnName(1));
    assertEquals(TEST_PROJECTION[2], projectionCursor.getColumnName(2));
    projectionCursor.close();

    projectionCursor = getMockContentResolver().query(
        TransactionProvider.ACCOUNTS_URI,
        TEST_PROJECTION,
        SELECTION_COLUMNS,
        SELECTION_ARGS,
        SORT_ORDER
    );

    assert projectionCursor != null;
    assertEquals(SELECTION_ARGS.length, projectionCursor.getCount());

    int index = 0;

    while (projectionCursor.moveToNext()) {

      assertEquals(SELECTION_ARGS[index], projectionCursor.getString(0));

      index++;
    }

    assertEquals(SELECTION_ARGS.length, index);
    projectionCursor.close();
  }

  public void testAccountIdUriWithMultipleBudgets() {
    long accountId = insertAccountWithTwoBudgets();

    Uri AccountIdUri = ContentUris.withAppendedId(TransactionProvider.ACCOUNTS_URI, accountId);

    Cursor cursor = getMockContentResolver().query(AccountIdUri,
        null,
        null,
        null,
        null
    );
    assert cursor != null;

    assertEquals(1, cursor.getCount());

    cursor.close();
  }

  public void testQueriesOnAccountIdUri() {
    final String SELECTION_COLUMNS = DatabaseConstants.KEY_LABEL + " = " + "?";

    final String query = "Account 0";
    final String[] SELECTION_ARGS = {query};

    final String[] Account_ID_PROJECTION = {
        DatabaseConstants.KEY_ROWID,
        DatabaseConstants.KEY_LABEL};

    insertData();

    Cursor cursor = getMockContentResolver().query(
        TransactionProvider.ACCOUNTS_URI,
        Account_ID_PROJECTION,
        null,
        null,
        null
    );
    assert cursor != null;

    assertEquals(TEST_ACCOUNTS.length, cursor.getCount());

    assertTrue(cursor.moveToFirst());

    int inputAccountId = cursor.getInt(0);

    Uri AccountIdUri = ContentUris.withAppendedId(TransactionProvider.ACCOUNTS_URI, inputAccountId);

    cursor.close();

    cursor = getMockContentResolver().query(AccountIdUri,
        new String[] {DatabaseConstants.KEY_LABEL},
        SELECTION_COLUMNS,
        SELECTION_ARGS,
        null
    );
    assert cursor != null;

    assertEquals(1, cursor.getCount());

    assertTrue(cursor.moveToFirst());

    assertEquals(query, cursor.getString(0));
    cursor.close();
  }

  public void testInserts() {
    Cursor cursor = getMockContentResolver().query(
        TransactionProvider.ACCOUNTS_URI,
        null,
        null,
        null,
        null
    );
    assert cursor != null;

    assertEquals(0, cursor.getCount());

    AccountInfo account = new AccountInfo(
        "Account 4",
        AccountType.ASSET, 1000);

    Uri rowUri = getMockContentResolver().insert(
        TransactionProvider.ACCOUNTS_URI,
        account.getContentValues()
    );

    long AccountId = ContentUris.parseId(rowUri);
    cursor.close();

    cursor = getMockContentResolver().query(
        TransactionProvider.ACCOUNTS_URI,
        null,
        null,
        null,
        null
    );
    assert cursor != null;

    assertEquals(1, cursor.getCount());

    assertTrue(cursor.moveToFirst());

    int descriptionIndex = cursor.getColumnIndex(DatabaseConstants.KEY_DESCRIPTION);
    int labelIndex = cursor.getColumnIndex(DatabaseConstants.KEY_LABEL);
    int balanceIndex = cursor.getColumnIndex(DatabaseConstants.KEY_OPENING_BALANCE);
    int currencyIndex = cursor.getColumnIndex(DatabaseConstants.KEY_CURRENCY);

    assertEquals(account.getLabel(), cursor.getString(labelIndex));
    assertEquals(account.getDescription(), cursor.getString(descriptionIndex));
    assertEquals(account.getOpeningBalance(), cursor.getLong(balanceIndex));
    assertEquals(account.getCurrency(), cursor.getString(currencyIndex));

    ContentValues values = account.getContentValues();

    values.put(DatabaseConstants.KEY_ROWID, AccountId);

    try {
      getMockContentResolver().insert(TransactionProvider.ACCOUNTS_URI, values);
      fail("Expected insert failure for existing record but insert succeeded.");
    } catch (Exception e) {

    }
    cursor.close();
  }

  public void testDeletes() {
    final String SELECTION_COLUMNS = DatabaseConstants.KEY_LABEL + " = " + "?";

    final String[] SELECTION_ARGS = {"Account 0"};

    int rowsDeleted = getMockContentResolver().delete(
        TransactionProvider.ACCOUNTS_URI,
        SELECTION_COLUMNS,
        SELECTION_ARGS
    );

    assertEquals(0, rowsDeleted);

    insertData();

    rowsDeleted = getMockContentResolver().delete(
        TransactionProvider.ACCOUNTS_URI,
        SELECTION_COLUMNS,
        SELECTION_ARGS
    );

    assertEquals(1, rowsDeleted);

    Cursor cursor = getMockContentResolver().query(
        TransactionProvider.ACCOUNTS_URI,
        null,
        SELECTION_COLUMNS,
        SELECTION_ARGS,
        null
    );
    assert cursor != null;

    assertEquals(0, cursor.getCount());
    cursor.close();
  }

  public void testUpdates() {

    final String SELECTION_COLUMNS = DatabaseConstants.KEY_LABEL + " = " + "?";

    final String[] selectionArgs = {"Account 1"};

    ContentValues values = new ContentValues();

    values.put(DatabaseConstants.KEY_LABEL, "Testing an update with this string");

    int rowsUpdated = getMockContentResolver().update(
        TransactionProvider.ACCOUNTS_URI,
        values,
        SELECTION_COLUMNS,
        selectionArgs
    );

    assertEquals(0, rowsUpdated);

    insertData();

    rowsUpdated = getMockContentResolver().update(
        TransactionProvider.ACCOUNTS_URI,
        values,
        SELECTION_COLUMNS,
        selectionArgs
    );

    assertEquals(1, rowsUpdated);
  }

  public void testGrouping() {
    final String[] Account_ID_PROJECTION = {
        DatabaseConstants.KEY_ROWID,
        DatabaseConstants.KEY_GROUPING
    };
    insertData();

    Cursor cursor = getMockContentResolver().query(
        TransactionProvider.ACCOUNTS_URI,
        Account_ID_PROJECTION,
        null,
        null,
        null
    );
    assert cursor != null;

    assertEquals(TEST_ACCOUNTS.length, cursor.getCount());

    assertTrue(cursor.moveToFirst());

    int inputAccountId = cursor.getInt(0);

    assertEquals(Grouping.NONE.name(), cursor.getString(1));

    Uri accountIdUri = ContentUris.withAppendedId(TransactionProvider.ACCOUNTS_URI, inputAccountId);

    cursor.close();

    getMockContentResolver().update(ContentUris.withAppendedId(TransactionProvider.ACCOUNT_GROUPINGS_URI, inputAccountId)
            .buildUpon().appendPath(Grouping.YEAR.name()).build(),
        null, null, null);


    cursor = getMockContentResolver().query(accountIdUri,
        new String[]{DbUtils.fqcn(DatabaseConstants.TABLE_ACCOUNTS, DatabaseConstants.KEY_GROUPING)},
        null,
        null,
        null
    );
    assert cursor != null;

    assertTrue(cursor.moveToFirst());

    assertEquals(Grouping.YEAR.name(), cursor.getString(0));
    cursor.close();
  }

  public void testSortDirection() {
    final String[] Account_ID_PROJECTION = {
        DatabaseConstants.KEY_ROWID,
        DatabaseConstants.KEY_SORT_DIRECTION
    };
    insertData();

    Cursor cursor = getMockContentResolver().query(
        TransactionProvider.ACCOUNTS_URI,
        Account_ID_PROJECTION,
        null,
        null,
        null
    );
    assert cursor != null;

    assertEquals(TEST_ACCOUNTS.length, cursor.getCount());

    assertTrue(cursor.moveToFirst());

    int inputAccountId = cursor.getInt(0);

    assertEquals(SortDirection.DESC.name(), cursor.getString(1));

    Uri accountIdUri = ContentUris.withAppendedId(TransactionProvider.ACCOUNTS_URI, inputAccountId);

    cursor.close();

    getMockContentResolver().update(accountIdUri.buildUpon().appendPath("sortDirection").appendPath(SortDirection.ASC.name()).build(),
        null, null, null);

    cursor = getMockContentResolver().query(accountIdUri,
        new String[] {DatabaseConstants.KEY_SORT_DIRECTION},
        null,
        null,
        null
    );
    assert cursor != null;

    assertTrue(cursor.moveToFirst());

    assertEquals(SortDirection.ASC.name(), cursor.getString(0));
    cursor.close();
  }
}

File: myExpenses/src/androidTest/java/org/totschnig/myexpenses/test/provider/AccountTest.kt
package org.totschnig.myexpenses.test.provider

import android.content.ContentResolver
import android.content.ContentUris
import android.content.ContentValues
import androidx.test.rule.provider.ProviderTestRule
import com.google.common.truth.Truth.assertThat
import org.junit.After
import org.junit.Assert.assertThrows
import org.junit.Rule
import org.junit.Test
import org.totschnig.myexpenses.model.Grouping
import org.totschnig.myexpenses.model.SortDirection
import org.totschnig.myexpenses.provider.AccountInfo
import org.totschnig.myexpenses.provider.BudgetInfo
import org.totschnig.myexpenses.provider.DatabaseConstants.KEY_AMOUNT
import org.totschnig.myexpenses.provider.DatabaseConstants.KEY_CURRENCY
import org.totschnig.myexpenses.provider.DatabaseConstants.KEY_DATE
import org.totschnig.myexpenses.provider.DatabaseConstants.KEY_DESCRIPTION
import org.totschnig.myexpenses.provider.DatabaseConstants.KEY_GROUPING
import org.totschnig.myexpenses.provider.DatabaseConstants.KEY_LABEL
import org.totschnig.myexpenses.provider.DatabaseConstants.KEY_OPENING_BALANCE
import org.totschnig.myexpenses.provider.DatabaseConstants.KEY_ROWID
import org.totschnig.myexpenses.provider.DatabaseConstants.KEY_SORT_BY
import org.totschnig.myexpenses.provider.DatabaseConstants.KEY_SORT_DIRECTION
import org.totschnig.myexpenses.provider.DatabaseConstants.KEY_SUM_EXPENSES
import org.totschnig.myexpenses.provider.DatabaseConstants.TABLE_ACCOUNTS
import org.totschnig.myexpenses.provider.DbUtils
import org.totschnig.myexpenses.provider.TransactionProvider
import org.totschnig.myexpenses.provider.TransactionProvider.SORT_URI
import org.totschnig.shared_test.CursorSubject.Companion.useAndAssert

class AccountTest {
    @get:Rule
    val providerRule: ProviderTestRule =
        ProviderTestRule.Builder(TransactionProvider::class.java, TransactionProvider.AUTHORITY)
            .build()

    private val resolver: ContentResolver
        get() = providerRule.resolver

    private val testAccounts = arrayOf(
        AccountInfo("Account 0", 1, 0),
        AccountInfo("Account 1", 2, 100),
        AccountInfo("Account 2", 3, -100)
    )

    private fun insertData() {
        //TODO set up account type
        for (account in testAccounts) {
            resolver.insert(
                TransactionProvider.ACCOUNTS_URI,
                account.contentValues
            )
        }
    }

    @After
    fun clearDb() {
        resolver.delete(
            TransactionProvider.ACCOUNTS_URI,
            null,
            null
        )
    }

    private fun insertAccountWithTwoBudgets(): Long {
        val grouping = Grouping.MONTH

        //TODO set up account type
        return ContentUris.parseId(
            resolver.insert(
                TransactionProvider.ACCOUNTS_URI,
                AccountInfo(
                    "Account with 2 budgets",
                    1,
                    -100,
                    "EUR"
                ).contentValues
            )!!
        ).also {
            resolver.update(
                ContentUris.withAppendedId(TransactionProvider.ACCOUNT_GROUPINGS_URI, it)
                    .buildUpon()
                    .appendPath(grouping.name).build(),
                null, null, null
            )
            val budgets = arrayOf(
                BudgetInfo(it, "budget 1", 400, grouping, "description"),
                BudgetInfo(it, "budget 2", 5000, grouping, "description")
            )
            for (budgetInfo in budgets) {
                resolver.insert(
                    TransactionProvider.BUDGETS_URI,
                    budgetInfo.contentValues
                )
            }
        }
    }

    @Test
    fun testQueriesOnAccountUri() {
        val testProjection = arrayOf(
            KEY_LABEL,
            KEY_DESCRIPTION,
            KEY_CURRENCY
        )
        val commentSelection = "$TABLE_ACCOUNTS.$KEY_LABEL = ?"
        val selectionColumns =
            "$commentSelection OR $commentSelection OR $commentSelection"
        val selectionArgs = arrayOf("Account 0", "Account 1", "Account 2")
        val sortOrder = "$KEY_LABEL ASC"
        resolver.query(
            TransactionProvider.ACCOUNTS_URI,
            null,
            null,
            null,
            null
        ).useAndAssert {
            hasCount(0)
        }
        insertData()

        resolver.query(
            TransactionProvider.ACCOUNTS_URI,
            null,
            null,
            null,
            null
        ).useAndAssert  {
            hasCount(testAccounts.size)
        }

        resolver.query(
            TransactionProvider.ACCOUNTS_URI,
            testProjection,
            null,
            null,
            null
        ).useAndAssert {
            hasColumnCount(testProjection.size)
            hasColumns(*testProjection)
        }

        resolver.query(
            TransactionProvider.ACCOUNTS_URI,
            testProjection,
            selectionColumns,
            selectionArgs,
            sortOrder
        ).useAndAssert {
            hasCount(selectionArgs.size)
            var index = 0
            while (actual.moveToNext()) {
                hasString(0, selectionArgs[index])
                index++
            }
        }
    }

    @Test
    fun testAccountIdUriWithMultipleBudgets() {
        val accountId = insertAccountWithTwoBudgets()
        val uri = ContentUris.withAppendedId(TransactionProvider.ACCOUNTS_URI, accountId)
        resolver.query(
            uri,
            null,
            null,
            null,
            null
        ).useAndAssert {
            hasCount(1)
        }
    }

    @Test
    fun testQueriesOnAccountIdUri() {
        val selection = "$TABLE_ACCOUNTS.$KEY_LABEL = ?"
        val query = "Account 0"
        val args = arrayOf(query)
        val projection = arrayOf(
            KEY_ROWID,
            KEY_LABEL
        )
        insertData()
        val inputAccountId = resolver.query(
            TransactionProvider.ACCOUNTS_URI,
            projection,
            null,
            null,
            null
        ).useAndAssert {
            hasCount(testAccounts.size)
            movesToFirst()
            actual.getLong(0)
        }
        val uri = ContentUris.withAppendedId(TransactionProvider.ACCOUNTS_URI, inputAccountId)

        resolver.query(
            uri, arrayOf(KEY_LABEL),
            selection,
            args,
            null
        ).useAndAssert {
            hasCount(1)
            movesToFirst()
            hasString(0, query)
        }
    }

    @Test
    fun testInserts() {
        resolver.query(
            TransactionProvider.ACCOUNTS_URI,
            null,
            null,
            null,
            null
        ).useAndAssert {
            hasCount(0)
        }
        val account = AccountInfo(
            "Account 4",
            4, 1000
        )

        val accountId = ContentUris.parseId(
            resolver.insert(
                TransactionProvider.ACCOUNTS_URI,
                account.contentValues
            )!!
        )

        resolver.query(
            TransactionProvider.ACCOUNTS_URI,
            null,
            null,
            null,
            null
        ).useAndAssert {
            val descriptionIndex = actual.getColumnIndex(KEY_DESCRIPTION)
            val labelIndex = actual.getColumnIndex(KEY_LABEL)
            val balanceIndex = actual.getColumnIndex(KEY_OPENING_BALANCE)
            val currencyIndex = actual.getColumnIndex(KEY_CURRENCY)
            hasCount(1)
            movesToFirst()
            hasString(labelIndex, account.label)
            hasString(descriptionIndex, account.description)
            hasLong(balanceIndex, account.openingBalance)
            hasString(currencyIndex, account.currency)
        }
        val values = account.contentValues
        values.put(KEY_ROWID, accountId)
        assertThrows(Exception::class.java) {
            resolver.insert(TransactionProvider.ACCOUNTS_URI, values)
        }
    }

    @Test
    fun testDeletes() {
        val selection = "$TABLE_ACCOUNTS.$KEY_LABEL = ?"
        val args = arrayOf("Account 0")
        assertThat(
            resolver.delete(
                TransactionProvider.ACCOUNTS_URI,
                selection,
                args
            )
        ).isEqualTo(0)
        insertData()
        assertThat(
            resolver.delete(
                TransactionProvider.ACCOUNTS_URI,
                selection,
                args
            )
        ).isEqualTo(1)
        resolver.query(
            TransactionProvider.ACCOUNTS_URI,
            null,
            selection,
            args,
            null
        ).useAndAssert {
            hasCount(0)
        }
    }

    @Test
    fun testUpdates() {
        val selection = "$TABLE_ACCOUNTS.$KEY_LABEL = ?"
        val selectionArgs = arrayOf("Account 1")
        val values = ContentValues().apply {
            put(KEY_LABEL, "Testing an update with this string")
        }

        assertThat(
            resolver.update(
                TransactionProvider.ACCOUNTS_URI,
                values,
                selection,
                selectionArgs
            )
        ).isEqualTo(0)
        insertData()
        assertThat(
            resolver.update(
                TransactionProvider.ACCOUNTS_URI,
                values,
                selection,
                selectionArgs
            )
        ).isEqualTo(1)
    }

    @Test
    fun testGrouping() {
        val projection = arrayOf(
            KEY_ROWID,
            KEY_GROUPING
        )
        insertData()
        val id = resolver.query(
            TransactionProvider.ACCOUNTS_URI,
            projection,
            null,
            null,
            null
        ).useAndAssert {
            hasCount(testAccounts.size)
            movesToFirst()
            hasString(1, Grouping.NONE.name)
            actual.getLong(0)
        }
        val uri = ContentUris.withAppendedId(TransactionProvider.ACCOUNTS_URI, id)

        resolver.update(
            ContentUris.withAppendedId(
                TransactionProvider.ACCOUNT_GROUPINGS_URI,
                id
            ).buildUpon().appendPath(Grouping.YEAR.name).build(),
            null, null, null
        )
        resolver.query(
            uri,
            arrayOf(DbUtils.fqcn(TABLE_ACCOUNTS, KEY_GROUPING)),
            null,
            null,
            null
        ).useAndAssert {
            movesToFirst()
            hasString(0, Grouping.YEAR.name)
        }
    }

    @Test
    fun testSortDirection() {
        val projection = arrayOf(
            KEY_ROWID,
            KEY_SORT_BY,
            KEY_SORT_DIRECTION
        )
        insertData()
        val id = resolver.query(
            TransactionProvider.ACCOUNTS_URI,
            projection,
            null,
            null,
            null
        ).useAndAssert {
            hasCount(testAccounts.size)
            movesToFirst()
            hasString(1, KEY_DATE)
            hasString(2, SortDirection.DESC.name)
            actual.getLong(0)
        }

        val accountIdUri =
            ContentUris.withAppendedId(TransactionProvider.ACCOUNTS_URI, id)

        resolver.update(
            ContentUris.withAppendedId(SORT_URI, id)
                .buildUpon()
                .appendPath(KEY_AMOUNT)
                .appendPath(SortDirection.ASC.name)
                .build(),
            null, null, null
        )
        resolver.query(
            accountIdUri, arrayOf(KEY_SORT_BY, KEY_SORT_DIRECTION),
            null,
            null,
            null
        ).useAndAssert {
            movesToFirst()
            hasString(0, KEY_AMOUNT)
            hasString(1, SortDirection.ASC.name)
        }
    }

    @Test
    fun testQueryWithSum() {
        val id = resolver.insert(
            TransactionProvider.ACCOUNTS_URI,
            AccountInfo("Account 0", 1, 0).contentValues
        ).let {
            ContentUris.parseId(it!!)
        }
        resolver.query(
            TransactionProvider.ACCOUNTS_FULL_URI,
            null,
            "$KEY_ROWID = ?",
            arrayOf(id.toString()),
            null
        )!!.useAndAssert {
            //TODO insert transactions and test calculated sum
            hasCount(1)
            movesToFirst()
            hasLong(KEY_SUM_EXPENSES, 0)
        }
    }
}